mongo/change_stream_test.go

```diff
@@ -173,6 +173,18 @@ func comparePipelines(t *testing.T, expectedraw, actualraw bson.Raw) {
 	}
 }
 
+// On some 4.0 driver tests running against more recent server versions, calls to ChangeStream.Next() loop indefinitely
+// because the server returns an empty batch of events for each getMore. This function inserts a document to the given
+// Collection to create an event for the stream to capture.
```

Comment:
The issue regarding error labels is different than the issue described here. The comments here are describing a case where there are no events for the change stream to return. The `Next` call indefinitely runs `getMore` commands, each of which succeeds but returns an empty batch of events. The issue described in COMPATIBILITY.md is about errors for `getMore`. In 4.0, drivers resumed on all errors except for 3 (Interrupted, CursorNotFound, and CappedPositionLost). We refined this approach in versions 4.2 and 4.4 to use error labels because there are more than 3 errors that should force the stream to stop. However, 4.0 drivers don't recognize these labels, so they continue to resume, which can result in an infinite loop.

Reply:
Ah ok, then why is this change only necessary on the 4.0 era driver? Don't we still (in the current version of the driver) call `getMore()` continuously when empty batches are returned?

Reply:
We do, but the testing infrastructure in the current version of the driver has drifted significantly from the 4.0 driver. We introduced the `mtest` framework, started creating new clients for each test, and put the integration tests in a new directory. The old infrastructure reused a bunch of clients and because the tests were in the `mongo` directory, tests would incorrectly modify client state. We found a number of bugs when we moved away from this approach.

Given how much this has diverged, I don't want to spend too much time figuring out why this issue with change streams only shows up in the 4.0 driver. We have a good understanding of the shortcomings of change streams in older driver versions as outlined in COMPATIBILITY.md. This is really the goal of the matrix testing project in the first place, so doing a deep dive to debug this exact issue that only occurs in tests on an unmaintained driver version seems unnecessary.

Reply:
Yeah that makes sense. It'll be interesting to see what kind of bugs come up from the matrix testing 🤔  Thanks!